### PR TITLE
ntfsck: add checking entry's vcn in ntfsck_validate_index_blocks()

### PR DESCRIPTION
**Pull request by jaysim**
When $INDEX_ALLOCATION were allocated, but not used (ie, $BITMAP cleared), ntfsck_validate_index_blocks() skip $INDEX_ALLOCATION block which is cleared on $BITMAP.

But, scanning directory tree will use ntfs_index_next(), that does not check $BITMAP.

So if index entries in $INDEX_ALLOCATION which is matched with cleared $BITMAP is corrupted, ntfsck cannot find it.

Add checking entry's vcn if it is set or not in $BITMAP. If not, it does not matched, so it initialize index entry.

### CODE REVIEW

File: src/ntfsck.c [hunk 1]
```diff
@@ -3144,14 +3183,39 @@ static void ntfsck_validate_index_blocks(ntfs_volume *vol,
...
 				ie = (INDEX_ENTRY *)((u8 *)ie + le16_to_cpu(ie->length))) {
-				/* Bounds checks. */
-				if (((u8 *)ie < (u8 *)ia) ||
+
+				/* check bitmap for sub-node */
+				if (ie->ie_flags & INDEX_ENTRY_NODE) {
+					VCN vcn = ntfs_ie_get_vcn(ie);
+					u32 sub_bmp_pos;
+
+					sub_bmp_pos = (vcn << ictx->vcn_size_bits) / ictx->block_size;
```
Comment (namjaejeon) [inline src/ntfsck.c]: we can move it to above if (!ntfs_bit_get(bmp_buf, sub_bmp_pos)).

File: src/ntfsck.c [hunk 2]
```diff
@@ -3144,14 +3183,39 @@ static void ntfsck_validate_index_blocks(ntfs_volume *vol,
 
 			for (; (u8 *)ie < index_end;
 				ie = (INDEX_ENTRY *)((u8 *)ie + le16_to_cpu(ie->length))) {
-				/* Bounds checks. */
-				if (((u8 *)ie < (u8 *)ia) ||
+
+				/* check bitmap for sub-node */
+				if (ie->ie_flags & INDEX_ENTRY_NODE) {
```
Comment (namjaejeon) [inline src/ntfsck.c:3201]: I wonder that this change should be under /* check length bound */ ?

File: src/ntfsck.c (hunk 1 above)
Comment (jaysim) [reply src/ntfsck.c]: I moved it above checking sub-node

File: src/ntfsck.c (hunk 2 above)
Comment (jaysim) [reply src/ntfsck.c:3201]: moved it under /* check length bound */

**Comment by namjaejeon** [comment]
@jaysim I have applied this patch after fixing the following checkpatch.pl warnings. Let me know if it is not good.

```
WARNING: quoted string split across lines
#49: FILE: src/ntfsck.c:3126:
+			ntfs_log_verbose("Index root entry out of bounds in"
+					" inode %"PRId64"\n", ni->mft_no);

WARNING: braces {} are not necessary for single statement blocks
#55: FILE: src/ntfsck.c:3132:
+			if (!ntfs_bit_get(bmp_buf, sub_bmp_pos)) {
+				goto initialize_index;
+			}

total: 0 errors, 2 warnings, 95 lines checked
```

**Pull request closed without merging**